syntax = "proto3";
option cc_enable_arenas = true;

package a3proto;
option java_package = "jp.gaje.analog3.protocol";

message SynthNodeCommon {
  enum NodeType {
    MODULE = 0;
    KNOB = 1;
    SWITCH = 2;
  }
  enum Channel {
    CAN = 0;
  }
  SynthNode node_type = 1;
  string node_name = 2;
  uint32 component_number = 3;
  Channel channel = 4;
  uint32 initial_value = 5;
}

message Module {
  uint32 model_id = 1;
}

message Knob {
  uint32 min_value = 1;
  uint32 max_value = 2;
}

message Switch {
  repeated string options = 1;
}

message SynthNode {
  SynthNodeCommon common = 1;
  // metadata
  Module module = 2;
  Knob knob = 3;
  Switch switch = 4;
  repeated SynthNode child_nodes = 5;
}

message SynthServiceMessage {
  enum Operation {
    PING                = 0;
    PING_RESP           = 1;
    LIST_MODELS         = 2;
    LIST_MODELS_RESP    = 3;
    GET_MODELS          = 4;
    GET_MODELS_RESP     = 5;
    LIST_INSTANCES      = 6;
    LIST_INSTANCES_RESP = 7;
    BIND                = 8;
    BIND_RESP           = 9;
    UNBIND              = 10;
    UNBIND_RESP         = 11;
    VALUE_CHANGE_REQ    = 12;
    VALUE_CHANGE_NOTIF  = 13;
  }
  Operation op                 = 1;
  uint32 sequence_num          = 2;
  repeated uint32 model_ids    = 3;
  repeated SynthNode models    = 4;
  repeated uint32 instance_ids = 5;
  uint32 instance_id           = 6;
  uint32 component_id          = 7;
  uint32 new_value             = 8;
}
